scripts/regenerate_piper_catalog.py

```
@@ -0,0 +1,379 @@
+        return max(0.0, delta)
+    except Exception:
+        return default
+
+
+def fetch_json(url, timeout=30, max_retries=6):
+    """GET JSON with exponential backoff on 429 (HF rate limit).
+
+    HF caps the tree API around 60 req/min; backoff is 2/4/8/16/32/64s
+    so a single endpoint can sleep up to ~2 minutes total before failing.
+    """
+    req = urllib.request.Request(_safe_url(url), headers={"Accept": "application/json"})
```

Ruff (0.15.11)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)